### ISSUE TITLE
Persistent Window Labels

### ISSUE THREAD

**Issue by gerases** [Hammerspoon/hammerspoon#1748]
Hello,

I use Chrome on a Mac. I also Tabs Outliner to break my windows into logical groups. One window consisting of 10 tabs could be called Ticket X. Another window could contain 20 more tabs relating to Ticket Y, etc.

Because many times windows look similar, I don't know right away what a window is really all about until look at a few tabs.

Is there a way to label my windows (containing multiple tabs) with persistent labels SOMEHOW? This would really boost my productivity.

Thanks!

**Comment by latenitefilms** [comment]
You could add a shortcut key to allow you to add an internal label to a window (this would just be stored internally in Hammerspoon - creating a link between the hs.window object and your description), then use a hs.chooser to switch between the windows?

**Comment by asmagill** [comment]
Haven't tested this, but this comes to mind: If you store the window object in a table with the label as they key, you can easily select the saved window, even if you're on a different space, and test if it's been closed by checking if its role is the empty string... (note that there are some limitations on what you can do with a window on another space, but focus *does* work by changing you to the space the window is currently on)

e.g.

~~~lua
-- global table to hole label-window mappings:
windowLabels = {}

-- for testing -- pick your labels however you choose
local myLabel = "Safari"

-- save a window with a label -- use whatever method you want to get the window object, I'm
-- just using the application finder and it's main window for demonstration purposes
windowLabels[myLabel] = hs.application("Safari"):mainWindow()

-- time passes

-- get the window saved with the label
local myWindow = windowLabels[myLabel]

if myWindow:role() ~= "" then
    myWindow:focus()
else
    hs.alert("Window or application has been closed")
    windowLabels[myLabel] = nil
end
~~~

**Comment by gerases** [comment]
Thank you for your replies. My first interest is in figuring out actually positioning the label. Maintaining the internal structure is important but secondary. Is there a hammerspoon plugin that upon a click on a window, will present the user with a dialog to the effect of "Add/Edit window label?" and then place it on the window (somewhere)?

**Comment by cmsj** [comment]
@gerases I don't believe such a plugin exists, but we do have hs.tabs which places a little button on windows - might be a good source of inspiration.

**Comment by gerases** [comment]
@cmsj thank you guys for your feedback. I found something that will replace this need. It's a plugin for chrome called Spaces. Thank you for your time and Hammerspoon!!

**Comment by runningskull** [comment]
Here's something I use in my config to get this functionality. You can name windows, and then switch to them by name using the chooser. I just hacked it together yesterday (caveat emptor), but am finding it super handy so far.

https://gist.github.com/runningskull/ddb2befbbcb3d8f75331c071827a2a70
